internal/datastore/mongo_store.go

```diff
@@ -114,14 +144,87 @@ func (mongodb MongoDB) DeleteQueue(id db.QueueId) error {
 	return nil
 }
 
-func (mongodb MongoDB) AddTokenToQueue(db.QueueId, db.Token) {
-	panic("Not implemented")
+func (mongodb MongoDB) AddTokenToQueue(id db.QueueId, token db.Token) (db.TokenId, error) {
+	token.Id = ""
+	token.QueueId = id
+
+	// Lock queue to ensure 2 concurrent add tokens don't happen
+	queueMutex.Lock()
+
+	// if there were 2 concurrent calls to this function, 2 tokens might get the same number
+	max, err := mongodb.GetMaxToken(id)
+
+	if err != nil {
+		log.Fatal(err)
+		queueMutex.Unlock()
+		return token.Id, err
+	}
+
+	token.TokenNumber = max + 1
+
+	result, err := mongodb.Token.InsertOne(context.TODO(), token)
+
+	queueMutex.Unlock()
+
+	if err != nil {
+		log.Println(err)
+		return token.Id, err
+	}
+
+	stringId := result.InsertedID.(primitive.ObjectID).Hex()
+
+	return db.TokenId(stringId), nil
 }
 
-func (mongodb MongoDB) ReadToken(db.TokenId) {
-	panic("Not implemented")
+func (mongodb MongoDB) GetMaxToken(id db.QueueId) (uint32, error) {
+	filter := bson.D{
+		// queueid is a string this time
+		{Key: "queueid", Value: id},
+	}
+
+	sort := bson.D{
+		{Key: "tokennumber", Value: -1},
```

Comment:
Guessing this is the initial value, let's set it to 0 so that the first token number becomes 1 instead of 0?

Reply:
No, this is just a sort filter. `-1` signifies descending order of sort. I'll add a comment to clarify that. Maybe change it to `sortFilter` instead.

Also, I am returning 0 if no tokens are found so that the first token number becomes 1.